collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java

```diff
@@ -0,0 +1,149 @@
+ * limitations under the License.
+ */
+
+package org.odk.collect.android;
+
+import android.content.res.AssetManager;
+import android.os.Environment;
+import android.support.test.InstrumentationRegistry;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.odk.collect.android.listeners.FormLoaderListener;
+import org.odk.collect.android.logic.FormController;
+import org.odk.collect.android.tasks.FormLoaderTask;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.concurrent.ExecutionException;
+
+import timber.log.Timber;
+
+import static junit.framework.Assert.assertEquals;
+
+/**
+ * This test has been created in order to check indices while navigating through a form.
+ * It's especially important while navigating through a form that contains nested groups and if we
+ * use groups with field-list appearance because in that case we need to collect all indices of
+ * questions we want to display on one page (we need to recursively get all indices contained in
+ * such a group and its children). It might be also tricky when navigating backwards because then we
+ * need to navigate to an index of the first question of all we want to display on one page.
+ */
+@RunWith(Parameterized.class)
+public class FormNavigationTestCase {
+
+    private static final String FORMS_DIRECTORY = "/odk/forms/";
+
+    @Parameters(name = "{0}")
+    public static Iterable<Object[]> data() {
+        // Expected indices when swiping forward until the end of the form and back once.
+        // An index of -1 indicates the start or end of a form
+        return Arrays.asList(new Object[][] {
+                {"simpleFieldList.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListInFieldList.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"regularGroupWithFieldListGroupInside.xml", new String[] {"-1, ", "0, 0, ", "-1, ", "0, 0, "}},
+                {"twoNestedRegularGroups.xml", new String[] {"-1, ", "0, 0, 0, ", "0, 0, 1, ", "0, 0, 2, ", "-1, ", "0, 0, 2, "}},
+                {"regularGroupWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 1, 1, ", "-1, ", "0, 1, 1, "}},
+                {"regularGroupWithQuestionsAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "}},
+                {"fieldListWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListWithQuestionsAndRegularGroupsInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"threeNestedFieldListGroups.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}}
+        });
+    }
```

Comment:
You might consider a further beautification like this, to remove some boilerplate and line things up better. If you prefer it how you have it now, that’s fine with me. This is inspired by domain-specific languages.
```
@Parameters(name = "{0}")
public static Iterable<Object[]> data() {
    // Expected indices when swiping forward until the end of the form and back once.
    // An index of -1 indicates the start or end of a form
    return Arrays.asList(
            ei("simpleFieldList.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("fieldListInFieldList.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("regularGroupWithFieldListGroupInside.xml",
                    "-1, ", "0, 0, ", "-1, ", "0, 0, "),
            ei("twoNestedRegularGroups.xml",
                    "-1, ", "0, 0, 0, ", "0, 0, 1, ", "0, 0, 2, ", "-1, ", "0, 0, 2, "),
            ei("regularGroupWithQuestionAndRegularGroupInside.xml",
                    "-1, ", "0, 0, ", "0, 1, 0, ", "0, 1, 1, ", "-1, ", "0, 1, 1, "),
            ei("regularGroupWithQuestionsAndRegularGroupInside.xml",
                    "-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "),
            ei("fieldListWithQuestionAndRegularGroupInside.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("fieldListWithQuestionsAndRegularGroupsInside.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("threeNestedFieldListGroups.xml",
                    "-1, ", "0, ", "-1, ", "0, "));
}

/** Expected indices for each form */
private static Object[] ei(String formName, String... expectedIndices) {
    return new Object[] {formName, expectedIndices};
}

```